pennylane/math/quantum.py

```diff
@@ -1093,7 +1093,7 @@ def _check_density_matrix(density_matrix):
             # Check if positive semi-definite
             evs, _ = qml.math.linalg.eigh(dm)
             evs = np.real(evs)
-            evs_non_negative = [ev for ev in evs if ev >= 0.0]
+            evs_non_negative = [ev for ev in evs if ev >= -1e-7]

```

Comment:
Technically, it's not by adding the trace distance, but this is something I realized when writing the tests for it. The problem with this line is that density matrices that are not full rank can have very small negative eigenvalues because of numerical inaccuracies.

For instance, take the following code:
```python
import pennylane as qml
from pennylane import numpy as np

dm = np.ones((4, 4)) / 4
print(np.linalg.eigh(dm)[0]
```
Here, `dm` is the density matrix of the uniform superposition over 2 qubits. It's a perfectly valid density matrix, but its eigenvalues are computed as `array([-2.47454167e-16, -8.56127406e-17, -3.08148791e-33,  1.00000000e+00])`. The presence of negative eigenvalues would then prevent it from being recognized as a valid density matrix.

As for why the value of `-1e-7`, that's kind of arbitrary, it has to be a small negative number, not too large to avoid false positive and not too small to still recognize valid density matrices.